### ISSUE TITLE
Ghosts introduced when unflagging flagsets after imaging with said flagsets

### ISSUE THREAD

**Issue by bennahugo** [ratt-ru/CubiCal#349]
According to the logs the dE solutions converges (de solutions having ~70%+ convergence on most tiles), but it seems to be converting on phase with opposite signs - the ghost is almost perfectly symmetric along the 45 diagonal, so roughly both l and m are flipped in sign for a narrowish field like this. 

Could this be the result of using chi^2 as a convergence criterion with wirtinger calculus instead of the traditional re, im separated solutions? 

I'm using the sr and fixing direction ```0``` and subtracting directions ```1:```. In this case there is only one dE direction, outlined by the region as indicated (left before de cal). The ghost indicated in yellow on the right:
![image](https://user-images.githubusercontent.com/7934586/74926054-f61a0680-53dd-11ea-87fd-924b91e82d6b.png)

I'm going to try peeling the source with CASA next to compare.

**Comment by bennahugo** [comment]
I should add that the rest of the field (within a 40 sigma mask) is only adjusted for dI frequency-independent phase.

**Comment by KshitijT** [comment]
![image](https://user-images.githubusercontent.com/9251962/74926616-d6371280-53de-11ea-8b41-fbd1df0a1c6c.png)

**Comment by bennahugo** [comment]
logs and parsets here:
ftp://elwood.ru.ac.za/pub/hugo/debug-logs/J1131-19551582125702.31.log
ftp://elwood.ru.ac.za/pub/hugo/debug-logs/J1131-19551582125702.31.parset

**Comment by bennahugo** [comment]
Solution bins 20x bigger makes no difference. As a final measure I will let it subtract out all the compact sources as a test

**Comment by bennahugo** [comment]
This time it is not an imager issue I should point out as a double check. The ghost appears in the dirty map after de cal after casa-based 2GC

**Comment by bennahugo** [comment]
The CASA loop that I will implement - bastardizing the cubical predictor to do the complicated uvsub below:

![image](https://user-images.githubusercontent.com/7934586/74960560-3a74c900-5415-11ea-827d-ee0524ff15a3.png)

This raises the disk space requirements 5 fold, but I don't think there is much of a choice. @IanHeywood is this roughly what you do for your peeling?

**Comment by bennahugo** [comment]
This should limit flux suppression I should add

**Comment by IanHeywood** [comment]
That looks the same as my understanding of what peeling is, although I haven't actually tried traditional peeling for years, having had differential gains generally cut the mustard without having to implement the teetering pile of scripts and scratch columns that peeling demands.

Maybe you're limited by the clean component model of the horrid region in the above example. Have you examined it to make sure it's garbage-free? Fitted components tend to work better for bright troublesome things in my experience. Stunts like over-sampled ~uniform-weighted images to get a high-resolution morphological model and suppress horrible sidelobes have mileage (or even consulting things like FIRST). 

The "noise" in the map also seems to have shot up after selfcal. Do the solutions look noisy? Are you sure a 40 sigma model is deep enough?

I think CASA's phase steering machinery uses small angle approximations so might not have the accuracy you require. The ft task also should never be used for predicting from CC models, but you already mentioned you're using CubiCal for that.

This looks like a tricky case, good luck. Would be keen to see your results. Cheers.

**Comment by bennahugo** [comment]
Thanks @IanHeywood I'm tracking this down to a bug in the model combination step in cubical, but I'm not yet any closer to figuring out what is the root cause behind it.

**Comment by bennahugo** [comment]
@IanHeywood it looks like it has problems attaching the data array. If you have previously run this with a python 2.7 version have you cleared caches?

**Comment by o-smirnov** [comment]
Ghost comment introduced and quickly deleted, @bennahugo? :)

**Comment by bennahugo** [comment]
Ok I've reproduced the pure imagingary components. They are all in flagged rows f = not(FLAG)!!!
```
CASA <23>: d = tb.getcol("MODEL_DATA")

CASA <24>: np.sum((d.real==0.0) & (d.imag!=0.0)) / float(d.size)
Out[24]: 0.29530839281501048

CASA <25>: d = tb.getcol("MODEL_OUT")

CASA <26>: np.sum((d.real==0.0) & (d.imag!=0.0)) / float(d.size)
Out[26]: 0.59061679717313709

CASA <27>: np.sum((d[f].real==0.0) & (d[f].imag!=0.0)) / float(d.size) # not flagged
Out[27]: 1.1543116078037005e-08

CASA <28>: d = tb.getcol("MODEL_DATA")

CASA <29>: np.sum((d[f].real==0.0) & (d[f].imag!=0.0)) / float(d.size) # not flagged
Out[29]: 0.0
```

So somehow cubical unflagged rows, exposing an invalid model. This means we have to be super duper careful not to run ddfacet (and / or wsclean?) after running cubical on a dataset. Fuck sake...

**Comment by bennahugo** [comment]
you can imagine this royally fucks up a workflow: 
- initial model using LEGACY
- running cubical SOLVER_FLAGS, 
- making a model with ddf using SOLVER_FLAGS
- rerunning cubical with cubical this time removing SOLVER_FLAGS and replacing them with a subset of SOLVER_FLAGS

This will introduce ghosts! Fucksake.... I'm taking this unholy power of cubical to roll to and through between flagsets out of my pipelines

**Comment by IanHeywood** [comment]
Yikes. I wonder if I've managed to get away with this by never letting CubiCal flag anything...

**Comment by IanHeywood** [comment]
Yikes. I wonder if I've managed to get away with this by never letting CubiCal flag anything...

**Comment by o-smirnov** [comment]
Well hey, running with powertools will occasionally lead to gory accidents... But settle down and try to understand the philosophy, @bennahugo, before you sow fear! Blindly stacking boolean flags has its own problems (flag creep, eventually leading to most of the data being flagged), so some flag management is required anyhow:

1. If you **repeat** a given workflow step, I would argue that it is imperative to first roll back the flags to their state before that step. Otherwise, your workflows become unreproducible. See also https://github.com/ska-sa/meerkathi/issues/760.

2. Solvers should be allowed to raise flags, this is generally considered a useful thing.

3. If you image after a solve, solver flags should be in effect. (As @JSKenyon pointed out and I fully agree, an imager's predict step ought to entirely fill the MODEL_DATA column regardless of any flag state, but neither DDF nor WSCLEAN seem to do that in their current incarnation, in an attempt to shave some runtime off. A case of being penny-wise and pound-foolish, in my opinion, but there we have it.)

4. If you solve again based on the updated MODEL_DATA, that's a **new** workflow step, not a repeated one, so the flags from that step should be saved under a different name.

The workflow you quote in https://github.com/ratt-ru/CubiCal/issues/349#issuecomment-590963352 is wrong because it neglects (4). If you run with ``--flags-apply legacy --flags-write cubical`` at the first step, and ``--flags-apply legacy,cubical --flags-write cubical2`` at the last step, the problem is avoided and the output flags are stacked as you expect... it's reusing the same flagset name at both steps that's twisted the flags out of shape.

Basically, CubiCal's bitflag workflow gives you a way to implement (1-4) without additional flag management steps in between. You can either embrace it (but then embrace it fully and use it the way it was intended), or avoid it entirely and use CASA's flagmanager *to do exactly the same job by hand*. I would argue the latter provides just as many opportunities for a fuck-up, but if you feel more comfortable with the more familiar (and, admittedly, better documented!) tool, that's perfectly fine. Either way, something in your workflow still needs to implement (1-4), or at least a similar flag management strategy!

**Comment by bennahugo** [comment]
Finally no ghosts and working dE cal by drawing boxes around sources... principle of maximal lazyness in action - after and before:
![image](https://user-images.githubusercontent.com/7934586/75345735-00d50f80-58a6-11ea-987b-137315ff9fcd.png)

**Comment by o-smirnov** [comment]
Now that is definitely a champagne-popping moment!

**Comment by o-smirnov** [comment]
OK I think this is happily resolved?
